gobase64.go

```diff
@@ -1,161 +1,87 @@
 package main
 
 import (
+	"encoding/binary"
 	"errors"
 	"io"
 	"log"
+	"math/bits"
 	"os"
 )
 
-func main() {
-	reader := os.Stdin
-	writer := os.Stdout
-
-	inbuf := make([]byte, 1024)
-	outbuf := make([]byte, 1368)
-
-	for {
-		octetsIn, err := reader.Read(inbuf)
-		if err == io.EOF {
-			break
-		} else if err != nil {
-			log.Fatal("Unexpected IO error", err)
-		}
-
-		n, err := Encode(&inbuf, octetsIn, &outbuf)
-
-		if err != nil {
-			log.Fatal("Unexpected Base64 encoding error", err)
-		}
-
-		writer.Write(outbuf[:n])
+// Not immutable but runtime array slicing to create a pseudo-immutable array is slower
+var charset = [64]byte{
+	'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
+	'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
+	'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1',
+	'2', '3', '4', '5', '6', '7', '8', '9', '+', '/'}
+
+// sextuple to chars
+// assumes output is initially empty
+func getChars(buffer *[]byte, padding int) (string, error) {
+	if padding > 2 {
+		return "", nil
 	}
-}
-
-func Encode(input *[]byte, octets int, output *[]byte) (int, error) {
-	triplets := octets / 3
-	remainder := octets % 3
-	var outlen int
-	if remainder == 0 {
-		outlen = triplets * 4
-	} else {
-		outlen = triplets*4 + 4
+	var out string
+	var in = *buffer
+	var l = len(in)
+	if l != 3 {
+		return "", errors.New("expected 3 bytes")
 	}
-
-	if outlen > len(*output) {
-		return 0, errors.New("output buffer is too small")
+	if padding > 0 {
+		in[2] = 0x00
 	}
-
-	in := *input
-	out := *output
-
-	// 1. read all triplets
-	for triplet := 0; triplet < triplets; triplet++ {
-		i := triplet * 3
-		j := triplet * 4
-		out[j] = toB64((in[i] & u6m) >> 2)
-		out[j+1] = toB64(((in[i] & l2m) << 4) | ((in[i+1] & u4m) >> 4))
-		out[j+2] = toB64(((in[i+1] & l4m) << 2) | ((in[i+2] & u2m) >> 6))
-		out[j+3] = toB64(in[i+2] & l6m)
+	if padding > 1 {
+		in[1] = 0x00
 	}
-
-	// 2. read remaining octets and add padding
-	if remainder != 0 {
-		i := triplets
-		j := i * 4
-
-		out[j] = toB64((in[i] & u6m) >> 2)
-
-		if remainder == 1 {
-			out[j+1] = toB64((in[i] & l2m) << 4) // zero-pad lower 4 bits
-			out[j+2] = pad
-			out[j+3] = pad
+	if l > 0 {
+		var unsigned = binary.BigEndian.Uint32(append([]byte{0x00}, in...))
+		var part1, part2, part3, part4 string
+		if padding > 0 {
+			part1 = ""
 		} else {
-			out[j+1] = toB64(((in[i] & l2m) << 4) | ((in[i+1] & u4m) >> 4))
-			out[j+2] = toB64((in[i+1] & l4m) << 2) // zero-pad lower 2 bits
-			out[j+3] = pad
+			part1 = string(charset[(unsigned & 0x003F)])
 		}
+		if padding > 1 {
+			part2 = ""
+		} else {
+			part2 = string(charset[bits.RotateLeft32((unsigned&0x0FC0), -6)])
+		}
+		part3 = string(charset[bits.RotateLeft32((unsigned&0x3F000), -12)])
+		part4 = string(charset[bits.RotateLeft32((unsigned&0xFC0000), -18)])
+		out = part4 + part3 + part2 + part1
```

Comment:
I like this approach of consing together the three octets of input. The masking / rotating ends up much more intelligible than what I tried with masking and bitwise-or'ing around byte boundaries.

Reply:
I'm not sure on using strings rather than reverting it back to bytes. I'm still digging into runes/strings/bytes for performance.  The bits package seems like best practice https://hermanschaaf.com/bit-manipulation-in-go-1-9/